File: CTF1/mysql.js
const mysql = require('mysql2/promise');

const dbConfig = {
  host: 'db-apache',
  user: 'app_user',
  password: 'app_password',
  database: 'web_app',
  charset: 'utf8mb4'
};

async function connectDb() {
  try {
    const connection = await mysql.createConnection(dbConfig);
    await connection.query("SET NAMES utf8mb4");
    await connection.query("SET CHARACTER SET utf8mb4");
    await connection.query("SET character_set_connection = utf8mb4");
    return connection;
  } catch (error) {
    console.error('Error de conexión:', error);
    return null;
  }
}

async function exec(sql, params = []) {
  const connection = await connectDb();
  if (!connection) return { rows: null, affectedRows: 0, insertId: undefined };
  try {
    const [rows] = await connection.execute(sql, params);
    const isHeader = rows && typeof rows.affectedRows === 'number';
    const affectedRows = isHeader ? rows.affectedRows : 0;
    const insertId = isHeader && rows.insertId ? rows.insertId : undefined;
    const dataRows = Array.isArray(rows) ? rows : [];
    return { rows: dataRows, affectedRows, insertId };
  } catch (error) {
    console.error('Error en la consulta:', error);
    return { rows: null, affectedRows: 0, insertId: undefined };
  } finally {
    try { await connection.end(); } catch {}
  }
}

async function queryDb(query, params = []) {
  const { rows } = await exec(query, params);
  return rows;
}

/* ------------------------- Repositorio: BLOGS ------------------------- */

async function listAllBlogs() {
  const sql = `SELECT * FROM blogs`;
  const rows = await queryDb(sql);
  return rows;
}


async function getBlogById(id) {
  const sql = `SELECT * FROM blogs WHERE id = ${id}`;
  const result = await exec(sql);
  return result.rows;
}

module.exports = {
  queryDb,
  listAllBlogs,
  getBlogById,
};

File: PythonWebApp/mysql.py
import mysql.connector

def connect_db():
    try:
        connection = mysql.connector.connect(
            host='db',
            user='app_user',
            password='app_password',
            database='web_app'
        )
        print('Conectado a la base de datos (MySQL)')
        return connection
    except Exception as e:
        print(f'Error de conexión: {e}')
        return None

def close_connection(connection):
    connection.close()
    print('Conexión cerrada')

def _build_sql(query, params):
    if params is None:
        return query
    if not isinstance(params, (list, tuple)):
        params = [params]
    for param in params:
        if param is None:
            value = 'NULL'
        elif isinstance(param, (int, float)):
            value = str(param)
        else:
            value = "'" + str(param).replace("'", "''") + "'"
        if '%s' in query:
            query = query.replace('%s', value, 1)
        elif '?' in query:
            query = query.replace('?', value, 1)
    return query

def query_db(query, params=None):
    connection = connect_db()
    if connection:
        cursor = connection.cursor(dictionary=True)
        try:
            sql = _build_sql(query, params)
            cursor.execute(sql)
            if query.strip().lower().startswith("select"):
                result = cursor.fetchall()
            else:
                result = cursor.rowcount
            connection.commit()
            return result
        except Exception as e:
            print(f'Error en la consulta: {e}')
            return None if query.strip().lower().startswith("select") else 0
        finally:
            cursor.close()
            close_connection(connection)